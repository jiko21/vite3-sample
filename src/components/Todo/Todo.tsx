import { css } from '@emotion/react';
import styled from '@emotion/styled';
import React, { useState } from 'react';
import { Todo as TodoType } from '../../todo';
import { TodoForm } from '../TodoForm';
import { TodoList } from '../TodoList';

const Wrapper = styled.div`
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
  padding: 0 20%;
`;

const Container = styled.div`
  width: 100%;
`;

const h2Style = css`
  color: #3c3c3c;
`;

export const Todo: React.VFC = () => {
  const [lists, setList] = useState<TodoType[]>([]);
  const [id, setId] = useState(0);

  const increment = () => {
    setId((state) => state + 1);
  };

  const handleSubmit = (text: string) => {
    setList((state) => [
      ...state,
      {
        id,
        content: text,
      },
    ]);
    increment();
  };

  const removeItem = (id: number) => {
    setList((state) => state.filter((item) => item.id !== id));
  };

  return (
    <Wrapper>
      <h2 css={h2Style}>Todo app</h2>
      <TodoForm handleSubmit={handleSubmit} />
      <Container>
        <TodoList todos={lists} onRemove={removeItem} testId="todo" />
      </Container>
    </Wrapper>
  );
};
