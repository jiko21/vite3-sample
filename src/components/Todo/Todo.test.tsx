import { afterEach, describe, expect, it, vi } from 'vitest';
import { Todo } from './Todo';
import { cleanup, fireEvent, render, screen } from '@testing-library/react';


describe('components/Todo', () => {
  afterEach(cleanup);

  it('correctly render', () => {
    const component = render(<Todo />);
    expect(component.container).toMatchSnapshot();
  });

  it('correctly add task', async () => {
    const component = render(<Todo />);
    const input = await screen.getByTestId('todoForm-input');
    fireEvent.change(input, {
      target: {
        value: 'test',
      },
    });
    const button = await screen.getByTestId('todoForm-button');
    fireEvent.click(button);
    expect(screen.getByText('test')).toBeTruthy();
    expect(component.container).toMatchSnapshot();
  });

  it('correctly not add task when input is empty', async () => {
    const component = render(<Todo />);
    const button = await screen.getByTestId('todoForm-button');
    fireEvent.click(button);
    expect(await screen.queryByText('remove')).toBeFalsy();
    expect(component.container).toMatchSnapshot();
  });

  it('correctly remove task', async () => {
    const component = render(<Todo />);
    const input = await screen.getByTestId('todoForm-input');
    const button = await screen.getByTestId('todoForm-button');
    fireEvent.change(input, {
      target: {
        value: 'test1',
      },
    });
    fireEvent.click(button);
    fireEvent.change(input, {
      target: {
        value: 'test2',
      },
    });
    fireEvent.click(button);
    fireEvent.change(input, {
      target: {
        value: 'test3',
      },
    });
    fireEvent.click(button);
    const removeButton = await screen.getByTestId('todo-remove-1-button');
    fireEvent.click(removeButton);
    expect(component.container).toMatchSnapshot();
    expect(await screen.queryByText('test1')).toBeTruthy();
    expect(await screen.queryByText('test2')).toBeFalsy();
    expect(await screen.queryByText('test3')).toBeTruthy();
  });

});
