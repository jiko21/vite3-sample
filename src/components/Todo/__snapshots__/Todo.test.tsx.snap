// Vitest Snapshot v1

exports[`components/Todo > correctly add task 1`] = `
<div>
  <div
    class="css-txwc29"
  >
    <h2
      class="css-l7qira-Todo"
    >
      Todo app
    </h2>
    <div>
      <input
        class="css-68qs4z"
        data-testid="todoForm-input"
        type="text"
        value=""
      />
      <button
        class="css-5shtp7-Button"
        data-testid="todoForm-button"
      >
        ADD
      </button>
    </div>
    <div
      class="css-1ago99h"
    >
      <ul
        class="css-gflyic"
      >
        <li
          class="css-ux9ziv"
        >
          <div
            class="css-1b7m767-Card"
          >
            <div
              class="css-1vo4ybp"
            >
              <span
                class="css-10rfslh"
              >
                test
              </span>
              <button
                class="css-gu30kj-Button"
                data-testid="todo-remove-0-button"
              >
                remove
              </button>
            </div>
          </div>
        </li>
      </ul>
    </div>
  </div>
</div>
`;

exports[`components/Todo > correctly not add task when input is empty 1`] = `
<div>
  <div
    class="css-txwc29"
  >
    <h2
      class="css-l7qira-Todo"
    >
      Todo app
    </h2>
    <div>
      <input
        class="css-68qs4z"
        data-testid="todoForm-input"
        type="text"
        value=""
      />
      <button
        class="css-5shtp7-Button"
        data-testid="todoForm-button"
      >
        ADD
      </button>
    </div>
    <div
      class="css-1ago99h"
    >
      <ul
        class="css-gflyic"
      />
    </div>
  </div>
</div>
`;

exports[`components/Todo > correctly remove task 1`] = `
<div>
  <div
    class="css-txwc29"
  >
    <h2
      class="css-l7qira-Todo"
    >
      Todo app
    </h2>
    <div>
      <input
        class="css-68qs4z"
        data-testid="todoForm-input"
        type="text"
        value=""
      />
      <button
        class="css-5shtp7-Button"
        data-testid="todoForm-button"
      >
        ADD
      </button>
    </div>
    <div
      class="css-1ago99h"
    >
      <ul
        class="css-gflyic"
      >
        <li
          class="css-ux9ziv"
        >
          <div
            class="css-1b7m767-Card"
          >
            <div
              class="css-1vo4ybp"
            >
              <span
                class="css-10rfslh"
              >
                test1
              </span>
              <button
                class="css-gu30kj-Button"
                data-testid="todo-remove-0-button"
              >
                remove
              </button>
            </div>
          </div>
        </li>
        <li
          class="css-ux9ziv"
        >
          <div
            class="css-1b7m767-Card"
          >
            <div
              class="css-1vo4ybp"
            >
              <span
                class="css-10rfslh"
              >
                test3
              </span>
              <button
                class="css-gu30kj-Button"
                data-testid="todo-remove-2-button"
              >
                remove
              </button>
            </div>
          </div>
        </li>
      </ul>
    </div>
  </div>
</div>
`;

exports[`components/Todo > correctly render 1`] = `
<div>
  <div
    class="css-txwc29"
  >
    <h2
      class="css-l7qira-Todo"
    >
      Todo app
    </h2>
    <div>
      <input
        class="css-68qs4z"
        data-testid="todoForm-input"
        type="text"
        value=""
      />
      <button
        class="css-5shtp7-Button"
        data-testid="todoForm-button"
      >
        ADD
      </button>
    </div>
    <div
      class="css-1ago99h"
    >
      <ul
        class="css-gflyic"
      />
    </div>
  </div>
</div>
`;
